fix: stop user code after 100ms of execution

ojeytonwilliams opened the pull request:
Code like `var xs = []; while(true){ xs.push(1) }` can quickly run the browser out of memory causing it to crash.  These changes use [@freecodecamp/loop-protect](https://www.npmjs.com/package/@freecodecamp/loop-protect) to break out of loops early (after a loop runs for more than 100ms) and warn the user that their code may contain an infinite loop.

`loop-protect` was also added back to the test runner as well, so that the user is protected when their code contains loops that are not evaluated until testing.  In testing the protection is weaker, waiting 2500ms before breaking out of a loop.

If this timeout proves to be too long, it can be lowered, potentially down to ~500ms.  In practice it's always going to be a compromise between good protection and allowing long running tests to complete.

Closes #37837

ojeytonwilliams committed:
fix: stop user code after 500ms of execution

Code like `var xs = []; while(true){ xs.push(1) }` can quickly run the
browser out of memory causing it to crash.  These changes stop user code
from executing for longer than 500ms in order to reduce the chance that
common mistakes will cause crashes.

RandellDawson commented:
I just tested the PR. However, even though I get see the error message in the console, about 750ms later, the browser still crashes.

![image](https://user-images.githubusercontent.com/5313213/69688865-215db580-107c-11ea-9a52-74065d716d98.png)

ojeytonwilliams committed:
fix: lower preview timeout to 100ms

ojeytonwilliams commented:
@freeCodeCamp/moderators if anyone could try this out and see if it stops their browser crashing, that would be a huge help.  It works for me, but not for Randell so I'd like to see if it's any use in practice.

Once you've got the PR running locally, just go to localhost:8000/learn/javascript-algorithms-and-data-structures/basic-javascript/iterate-with-javascript-while-loops (using Chrome) and enter

```js
var myArray = [];

var i = 0;
while(i < 5) {
  myArray.push(i);
  i--;
}
```

It should output to the fcc console and *not* crash, but that's not guaranteed!

raisedadead commented on client/src/templates/Challenges/redux/execute-challenge-saga.js:
```diff
@@ -163,10 +167,14 @@ function* previewChallengeSaga() {
   } catch (err) {
+    if (err === 'timeout') {
+      // eslint-disable-next-line no-ex-assign
+      err = `The code you have written is taking longer than the ${previewTimeout}ms our challenges allow. You may have created an infinite loop or need to rewrite an inefficient algorithm`;
```
```suggestion
      err = `The code you have written is taking longer than the ${previewTimeout}ms our challenges allow. You may have created an infinite loop or need to write an efficient algorithm`;
```

Let's change the double negative verbiage.

ojeytonwilliams committed:
Update client/src/templates/Challenges/redux/execute-challenge-saga.js

Co-Authored-By: mrugesh <1884376+raisedadead@users.noreply.github.com>

moT01 commented:
I mentioned in the issue that I think chrome throws the error or whatever after 3600ms - 100ms seems a little short to me. I would rather give it like 2500ms.

ojeytonwilliams commented:
@moT01 I'd like to have it be longer, but it already crashes my machine intermittently at 500ms and Randell's still crashes at 100ms.

ojeytonwilliams committed:
fix: update and use loop-protect for js preview

This informs the user and stops execution if a long running loop is
detected (js and jsx challenges) during preview.  Before this there was
no protection for js challenges and no information was given to the
user if they had created such a loop.

JavaScript tests are not affected (needed because some valid solutions
are also slow), but JSX tests now fail if a very slow loop is detected.

The latest published version is 2.1.6, but that does not include the
changes needed to use Babel >= 7, hence the direct Github link.

ojeytonwilliams commented:
This could have been a bit simpler, but I wanted to allow JS tests to run for a long time (that was the original motivation for the workers, after all) while using loop-protect to keep the previews from crashing.

@moT01 once I used loop-protect, I was able to raise the preview timer to 2500ms without any obvious issues.

ojeytonwilliams committed:
fix: report loops rather than throw errors

Now the js and jsx preview continues after infinite loops, but users are
warned that the tests may fail.

RandellDawson commented on client/src/templates/Challenges/rechallenge/transformers.js:
```diff
@@ -23,7 +23,14 @@ import createWorker from '../utils/worker-executor';
+
+function loopProtectCB(line) {
+  console.log(
+    `Potentially infinite loop detected on line ${line}. Tests may fail if this is not changed`
```
```suggestion
    `Potentially infinite loop detected on line ${line}. Tests may fail if this is not changed.`
```

ojeytonwilliams committed:
Update client/src/templates/Challenges/rechallenge/transformers.js

Co-Authored-By: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

raisedadead commented:
LGTM, I am not a fan of depending on GitHub for the package. Can we fork and publish the version we need @ojeytonwilliams ? If yes, I can take care of it.

ojeytonwilliams commented:
@raisedadead do you mean a scoped version like `@freecodecamp/loop-protect`?  I assume so, it uses the MIT license.

After discussing with @RandellDawson I added loop-protect back into the js tests, but with a long timeout to avoid the problems mentioned in https://github.com/freeCodeCamp/freeCodeCamp/pull/34524.  This should make it less likely that the tests crash the browser, though we might need to tune the precise timeout later.

ojeytonwilliams commented:
I've created a [fork](https://github.com/ojeytonwilliams/loop-protect) of `loop-protect` and created the necessary changes on that.  We'll need to tune the numbers so that it's as restrictive as possible while still allowing the slow challenges to complete.

ojeytonwilliams committed:
feat: use more performant loop-protect for tests

ojeytonwilliams commented:
The latest commit is set so that, on my machine, it doesn't crash (barely) if you try to use

```js
var xs = [];
while(true) {
  xs.push(i++);
}
```
While still passing `npm run test:curriculum`.  There is not much leeway.  Increasing the timeout much more than 4000ms crashes my browser and anything below 2500ms does not reliably pass the tests.

Assuming this approach is okay, I think we should try lowering the timeout and relax any tests that fail, if possible and if the solution is a reasonable algorithm.

raisedadead commented:
I was wondering if we could hook into some other API for the render. For instance [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) could be used to check if the test runner (worker) crashed and display a message, if not proceed to render.

ojeytonwilliams commented:
I'm not sure that will help.  The code is already evaluated inside a try block

```js
    try {
      testResult = eval(`
        ${e.data.build}
        __userCodeWasExecuted = true;
        ${e.data.testString}
      `);
    } catch (err) {
```

and any actual errors are handled nicely.  In these cases I think the JS engine is going OOM and the entire browser has to shut down as a result, so the only thing we can do is try to avoid them.

I mean, if you think it's worth a try, I can see if it works.  I'm not optimistic, though.

raisedadead commented:
Hmm, I see. My intent is to be able to separate the render loop from the execution loop. (The workers were intended to do this)

ojeytonwilliams committed:
fix: use @freecodecamp/loop-protect

Also changed the config, slightly since it can still crash.

ojeytonwilliams commented:
I've published the fork and updated the reference.  When I checked it locally it still crashed at 3000ms, so I lowered it slightly.  This may result in the tests occasionally failing, so, when I can, I'll modify those challenges.

ojeytonwilliams commented:
Blocked until the tests pass reliably.

ojeytonwilliams committed:
fix: speed up tests

Either improved the algorithm or reduced the difficulty for the slowest
tests. All these pass with testProtectTimeout = 500, so should be safe
on most machines with the current timeout of 2500.

ojeytonwilliams commented:
I think, pending QA, this is in good shape.  The lowest I went in testing was 250ms, which resulted in at least one project Euler challenge failing.  At 500ms everything worked just fine.

raisedadead commented:
@moT01 @RandellDawson do you mind taking a look?

moT01 commented on client/src/templates/Challenges/rechallenge/transformers.js:
```diff
@@ -23,17 +23,42 @@ import createWorker from '../utils/worker-executor';
+
+function loopProtectCB(line) {
+  console.log(
+    `Potentially infinite loop detected on line ${line}. Tests may fail if this is not changed.`
```
```suggestion
    `Potential infinite loop detected on line ${line}. Tests may fail if this is not changed.`
```

moT01 replied:
The grammar here doesn't sound right to me - not quite sure if it is or not. I'll let you decide.

ojeytonwilliams replied:
I *slightly* prefer potentially, but I could be swayed.  @RandellDawson do you have a preference?

raisedadead replied:
I am inclined on Tom's suggestion i.e potential

thecodingaviator replied:
`Potential` sounds better to me

moT01 commented on client/src/templates/Challenges/rechallenge/transformers.js:
```diff
@@ -23,17 +23,42 @@ import createWorker from '../utils/worker-executor';
+
+function testLoopProtectCB(line) {
+  console.log(
+    `Potentially infinite loop detected on line ${line}. Tests may be failing because of this.`
```
```suggestion
    `Potential infinite loop detected on line ${line}. Tests may be failing because of this.`
```

moT01 commented on client/src/templates/Challenges/redux/execute-challenge-saga.js:
```diff
@@ -155,18 +158,22 @@ function* previewChallengeSaga() {
   } catch (err) {
+    if (err === 'timeout') {
+      // eslint-disable-next-line no-ex-assign
+      err = `The code you have written is taking longer than the ${previewTimeout}ms our challenges allow. You may have created an infinite loop or need to write an efficient algorithm`;
```
```suggestion
      err = `The code you have written is taking longer than the ${previewTimeout}ms our challenges allow. You may have created an infinite loop or need to write a more efficient algorithm`;
```

moT01 replied:
Minor suggested change here, I think this sounds a little friendlier.

moT01 reviewed (commented):
A few minor grammar suggestions for now.

Other than that, I could not get my chrome to throw that error.

I did not dive into the code here yet - I think @RandellDawson should see if his browser is having any issues since he was when you (@ojeytonwilliams ) were not, and if it seems to be working for him, I can try and give the code a better review.

ojeytonwilliams committed:
Update client/src/templates/Challenges/redux/execute-challenge-saga.js

Co-Authored-By: Tom <20648924+moT01@users.noreply.github.com>

ojeytonwilliams commented:
> Other than that, I could not get my chrome to throw that error.

Does it throw on live?

moT01 commented:
Yes, Chrome throws the error on production for me - The "Aw, Snap!" screen shows up.

ojeytonwilliams committed:
Update client/src/templates/Challenges/rechallenge/transformers.js

Improve grammar

Co-Authored-By: Tom <20648924+moT01@users.noreply.github.com>

ojeytonwilliams committed:
Update client/src/templates/Challenges/rechallenge/transformers.js

Co-Authored-By: Tom <20648924+moT01@users.noreply.github.com>

raisedadead commented:
I pulled this down and tested, and seems to be working fine for the challenge in the linked issue. Should we merge this and wait for more feedback?

raisedadead committed:
Merge branch 'master' into fix/lower-preview-timeout

ojeytonwilliams commented:
I'd like to wait until @RandellDawson has had a chance to test it, since his machine seemed to crash the most readily.

raisedadead commented:
Sure thing.

RandellDawson commented:
Sorry I have been attending to other stuff.  I will look at this today.

iansibindi commented:
Hy
On 16 Dec 2019 17:33, "Randell Dawson" <notifications@github.com> wrote:

> Sorry I have been attending to other stuff. I will look at this today.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/freeCodeCamp/freeCodeCamp/pull/37841?email_source=notifications&email_token=AF7VISH5X35HFSOM4TIIIKLQY6GSXA5CNFSM4JSAHO7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG64F4Y#issuecomment-566084339>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF7VISGPACBBBKVHPA6IJKLQY6GSXANCNFSM4JSAHO7A>
> .
>

ojeytonwilliams committed:
fix: lower loop-protect timeout

RandellDawson commented:
@raisedadead This PR was not ready for merging.  I have been working with @ojeytonwilliams to test it and things were still not working properly on my machine.

raisedadead commented:
I wanted to test this out on the .dev, from what I know its not going to make the experience worser? From what I understand it will still be resulting in what we have on production for some machines (like yours)?

raisedadead commented:
Feel free to report fresh issues as you find them.

RandellDawson commented:
> From what I understand it will still be resulting in what we have on production for some machines (like yours)?

It does not report to the user an infinite loop in the FCC console like it is supposed to.  It just says tests timed out.

ojeytonwilliams commented:
Js challenges on .dev are completely broken right now:
![image](https://user-images.githubusercontent.com/15801806/71300452-d68a4180-238c-11ea-9f78-9b16a7ab7860.png)

almost certainly due to my PR.  I'm not, yet, sure why since it works perfectly on my machine
